src/overload_hessian.jl

```diff
@@ -1,85 +1,206 @@
             else
                 return t
             end
         else
-            return T(t.grad, t.hess ∪ (t.grad × t.grad))
+            return T(gradient(t), hessian(t) ∪ (gradient(t) × gradient(t)))
+        end
+    end
+    @eval function Base.$fn(t::D) where {P,T<:HessianTracer,D<:Dual{P,T}}
+        x = primal(t)
+        out = Base.$fn(x)
+        if is_seconder_zero_local($fn, x)
+            if is_firstder_zero_local($fn, x)
+                return Dual(out, empty(T))
+            else
+                return Dual(out, t)
+            end
+        else
+            return Dual(out, T(gradient(t), hessian(t) ∪ (gradient(t) × gradient(t))))
         end
     end
 end
 
 ## 2-to-1
 for fn in ops_2_to_1
-    @eval function Base.$fn(a::T, b::T) where {G,H,T<:GlobalHessianTracer{G,H}}
+    @eval function Base.$fn(a::T, b::T) where {G,H,T<:HessianTracer{G,H}}
         grad = empty(G)
         hess = empty(H)
         if !is_firstder_arg1_zero_global($fn)
-            grad = union(grad, a.grad) # TODO: use union!
-            union!(hess, a.hess)
+            grad = union(grad, gradient(a)) # TODO: use union!
+            union!(hess, hessian(a))
         end
         if !is_firstder_arg2_zero_global($fn)
-            grad = union(grad, b.grad) # TODO: use union!
-            union!(hess, b.hess)
+            grad = union(grad, gradient(b)) # TODO: use union!
+            union!(hess, hessian(b))
         end
         if !is_seconder_arg1_zero_global($fn)
-            union!(hess, a.grad × a.grad)
+            union!(hess, gradient(a) × gradient(a))
         end
         if !is_seconder_arg2_zero_global($fn)
-            union!(hess, b.grad × b.grad)
+            union!(hess, gradient(b) × gradient(b))
         end
         if !is_crossder_zero_global($fn)
-            union!(hess, (a.grad × b.grad) ∪ (b.grad × a.grad))
+            union!(hess, (gradient(a) × gradient(b)) ∪ (gradient(b) × gradient(a)))
         end
         return T(grad, hess)
     end
+    @eval function Base.$fn(a::D, b::D) where {P,G,H,T<:HessianTracer{G,H},D<:Dual{P,T}}
+        x = primal(a)
+        y = primal(b)
+        out = Base.$fn(x, y)
+
+        grad = empty(G)
+        hess = empty(H)
+        if !is_firstder_arg1_zero_local($fn, x, y)
+            grad = union(grad, gradient(a)) # TODO: use union!
+            union!(hess, hessian(a))
+        end
+        if !is_firstder_arg2_zero_local($fn, x, y)
+            grad = union(grad, gradient(b)) # TODO: use union!
+            union!(hess, hessian(b))
+        end
+        if !is_seconder_arg1_zero_local($fn, x, y)
+            union!(hess, gradient(a) × gradient(a))
+        end
+        if !is_seconder_arg2_zero_local($fn, x, y)
+            union!(hess, gradient(b) × gradient(b))
+        end
+        if !is_crossder_zero_local($fn, x, y)
+            union!(hess, (gradient(a) × gradient(b)) ∪ (gradient(b) × gradient(a)))
+        end
+        return Dual(out, T(grad, hess))
```

Comment:
Everything in this function is identical to the one before, except the last line.
Maybe treatment of ordinary tracers and duals could be unified for the tracing part, and only at the end do you decide what you return?

Reply:
I know there is a difference between `der_zero_local(op, x, y)` and `der_zero_global(op)`, but maybe `primal(xt)` could return `nothing` on a simple `GradientTracer` or something